ctags/main/entry.c

```diff
@@ -377,38 +508,948 @@ static int replacementTruncate (const char *const name, const long size)
+	if (slot->extensionFields.varType)
+		eFree ((char *)slot->extensionFields.varType);
+/* GEANY DIFF END */
+#ifdef HAVE_LIBXML
+	if (slot->extensionFields.xpath)
+		eFree ((char *)slot->extensionFields.xpath);
+#endif
+
+	if (slot->sourceLanguage)
+		eFree ((char *)slot->sourceLanguage);
+	if (slot->sourceFileName)
+		eFree ((char *)slot->sourceFileName);
+
+	clearParserFields (slot);
+}
+
+static unsigned int queueTagEntry(const tagEntryInfo *const tag)
+{
+	unsigned int i;
+	void *tmp;
+	tagEntryInfo * slot;
+
+	if (! (TagFile.corkQueue.count < TagFile.corkQueue.length))
+	{
+		if (!TagFile.corkQueue.length)
+			TagFile.corkQueue.length = 1;
+
+		tmp = eRealloc (TagFile.corkQueue.queue,
+				sizeof (*TagFile.corkQueue.queue) * TagFile.corkQueue.length * 2);
+
+		TagFile.corkQueue.length *= 2;
+		TagFile.corkQueue.queue = tmp;
+	}
+
+	i = TagFile.corkQueue.count;
+	TagFile.corkQueue.count++;
+
+
+	slot = TagFile.corkQueue.queue + i;
+	recordTagEntryInQueue (tag, slot);
+
+	return i;
+}
+
+
+static void *writerData;
+static tagWriter *writer;
+
+extern void setTagWriter (tagWriter *t)
+{
+	writer = t;
+}
+
+extern bool outpuFormatUsedStdoutByDefault (void)
+{
+	return writer->useStdoutByDefault;
+}
+
+extern void setupWriter (void)
+{
+	if (writer->preWriteEntry)
+		writerData = writer->preWriteEntry (TagFile.mio);
+	else
+		writerData = NULL;
+}
+
+extern void teardownWriter (const char *filename)
+{
+	if (writer->postWriteEntry)
+		writer->postWriteEntry (TagFile.mio, filename, writerData);
+}
+
+static void buildFqTagCache (const tagEntryInfo *const tag)
+{
+	renderFieldEscaped (FIELD_SCOPE_KIND_LONG, tag, NO_PARSER_FIELD);
+	renderFieldEscaped (FIELD_SCOPE, tag, NO_PARSER_FIELD);
+}
+
+#ifdef CTAGS_LIB
+static void initCtagsTag(ctagsTag *tag, const tagEntryInfo *info)
+{
+	tag->name = info->name;
+	tag->signature = info->extensionFields.signature;
+	tag->scopeName = info->extensionFields.scopeName;
+	tag->inheritance = info->extensionFields.inheritance;
+	tag->varType = info->extensionFields.varType;
+	tag->access = info->extensionFields.access;
+	tag->implementation = info->extensionFields.implementation;
+	tag->kindLetter = info->kind->letter;
+	tag->isFileScope = info->isFileScope;
+	tag->lineNumber = info->lineNumber;
+}
+#endif
+
+static void writeTagEntry (const tagEntryInfo *const tag)
+{
+	int length = 0;
+
+	if (tag->placeholder)
+		return;
+	if (! tag->kind->enabled)
+		return;
+	if (tag->extensionFields.roleIndex != ROLE_INDEX_DEFINITION
+	    && ! isXtagEnabled (XTAG_REFERENCE_TAGS))
+		return;
+
+	DebugStatement ( debugEntry (tag); )
+	Assert (writer);
+
+	if (includeExtensionFlags ()
+	    && isXtagEnabled (XTAG_QUALIFIED_TAGS)
+	    && doesInputLanguageRequestAutomaticFQTag ())
+		buildFqTagCache (tag);
+
+#ifdef CTAGS_LIB
+	getTagScopeInformation((tagEntryInfo *)tag, NULL, NULL);
+
+	if (TagEntryFunction != NULL)
+	{
+		ctagsTag t;
+
+		initCtagsTag(&t, tag);
+		length = TagEntryFunction(&t, TagEntryUserData);
+	}
+#else
```

Comment:
This is something I was considering too. We could "write" tags into Geany's structures in the callback of the writer and use normal way of ctags initialization and parsing. 

What however worries me is the MIO (TagFile.mio from entry.c) into which a writer is supposed to write - right now it's automatically opened/closed by ctags for any writer and that's not something we want. If we somehow avoid it and decide to keep it NULL, we have to be very careful that nothing from entry.c that refers it. That would be quite a big change and would make the diff bigger against uctags (and more fragile when something upstream changes).

I agree this is probably the right way it should be done and this is also the way it should probably be introduced into uctags. But I think having this shortcut in Geany which makes sure nothing with unexpected side-effects from ctags is called is better for now.